Release tools: detect and handle the 409 error

## 📝 Provide a description of the new feature

Sometimes, npm reports the following error while publishing the nightly packages.

```
npm notice === Tarball Details === 
npm notice name:          @ckeditor/ckeditor5-heading                            
npm notice version:       0.0.0-nightly-20240626.0                               
npm notice filename:      ckeditor-ckeditor5-heading-0.0.0-nightly-20240626.0.tgz         
...
npm notice Publishing to https://registry.npmjs.org/ with tag nightly and public access
npm ERR! code E409
npm ERR! 409 Conflict - PUT https://registry.npmjs.org/@ckeditor%2fckeditor5-heading - Failed to save packument. A common cause is if you try to publish a new package before the previous package has been fully processed.                     
```

The [`publishPackages()`](https://github.com/ckeditor/ckeditor5-dev/blob/master/packages/ckeditor5-dev-release-tools/lib/tasks/publishpackages.js) task rejected a promise when executing the `npm publish` command. After retrying execution several times ([ref](https://github.com/ckeditor/ckeditor5/blob/4a4dd185902f98eb29d56eabd959c686c6fd791d/scripts/release/publishpackages.js#L82)), CI publishes the same package thrice. It fails as the first call did the job even if the promise was rejected.

We need to be more innovative. Let's detect the 409 error suggesting publishing the same package twice.

### Definition of Done

- If a package lands on npm, do not publish it again. Instead, the task should resolve a promise.

### Technical details

- When `npm publish` fails, add a package to a stack and continue.
- Once all packages are processed, review the stack.
- For each package from the stack, see if it's published.
	```
	npm show @ckeditor/ckeditor5-heading@0.0.0-nightly-20240626.0
	
	@ckeditor/ckeditor5-heading@0.0.0-nightly-20240626.0 | GPL-2.0-or-later | deps: 1 | versions: 506
	```
- If so, mark the package as processed successfully.
- Otherwise, reject the promise.

### Other details

_No response_

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

--- a/scripts/release/publishpackages.js
+++ b/scripts/release/publishpackages.js
@@ -82,6 +82,15 @@ const tasks = new Listr( [
 		retry: 3
 	},
 	{
+		title: 'Checking if packages that returned 409 errors were uploaded correctly.',
+		task: async () => {
+			return releaseTools.verifyPackagesPublishedCorrectly( {
+				packagesDirectory: RELEASE_NPM_DIRECTORY,
+				version: latestVersion
+			} );
+		}
+	},
+	{
 		title: 'Pushing changes.',
 		task: () => {
 			return releaseTools.push( {

--- a/scripts/release/publishpackages.js
+++ b/scripts/release/publishpackages.js
@@ -82,7 +82,7 @@ const tasks = new Listr( [
 		retry: 3
 	},
 	{
-		title: 'Checking if packages that returned 409 errors were uploaded correctly.',
+		title: 'Checking if packages that returned E409 error code were uploaded correctly.',
 		task: async ( _, task ) => {
 			return releaseTools.verifyPackagesPublishedCorrectly( {
 				packagesDirectory: RELEASE_NPM_DIRECTORY,

--- a/scripts/release/publishpackages.js
+++ b/scripts/release/publishpackages.js
@@ -83,10 +83,13 @@ const tasks = new Listr( [
 	},
 	{
 		title: 'Checking if packages that returned 409 errors were uploaded correctly.',
-		task: async () => {
+		task: async ( _, task ) => {
 			return releaseTools.verifyPackagesPublishedCorrectly( {
 				packagesDirectory: RELEASE_NPM_DIRECTORY,
-				version: latestVersion
+				version: latestVersion,
+				onSuccess: text => {
+					task.output = text;
+				}
 			} );
 		}
 	},

